cloudconfig/templates.go

```diff
@@ -72,22 +72,26 @@ if [ ! -z $GH_RUNNER_GROUP ];then
     RUNNER_GROUP_OPT="--runnergroup=$GH_RUNNER_GROUP"
 fi
 
+# test if runner is already installed
```

Comment:
Let's try something like this:

```bash
# This will echo the version number in the filename. Given a file name like: actions-runner-osx-x64-2.299.1.tar.gz
# this will output: 2.299.1
function getRunnerVersion() {
    FILENAME="{{ .FileName }}"
    [[ $FILENAME =~ ([0-9]+\.[0-9]+\.[0-9+]) ]]
    echo $BASH_REMATCH
}

function getCachedToolsPath() {
    VERSION=$(getRunnerVersion)
    if [ -z "$VERSION" ]; then
        return 0
    fi

    CACHED_RUNNER="/opt/cache/actions-runner/$VERSION"
    if [ -d "$CACHED_RUNNER" ];then
        echo "$CACHED_RUNNER"
        return 0
    fi
    return 0
}

function downloadAndExtractRunner() {
    if [ ! -z "{{ .TempDownloadToken }}" ]; then
	TEMP_TOKEN="Authorization: Bearer {{ .TempDownloadToken }}"
    fi
    curl -L -H "${TEMP_TOKEN}" -o "/home/{{ .RunnerUsername }}/{{ .FileName }}" "{{ .DownloadURL }}" || fail "failed to download tools"
    mkdir -p /home/runner/actions-runner || fail "failed to create actions-runner folder"
    sendStatus "extracting runner"
    tar xf "/home/{{ .RunnerUsername }}/{{ .FileName }}" -C /home/{{ .RunnerUsername }}/actions-runner/ || fail "failed to extract runner"
    chown {{ .RunnerUsername }}:{{ .RunnerGroup }} -R /home/{{ .RunnerUsername }}/actions-runner/ || fail "failed to change owner"
}

CACHED_RUNNER=$(getCachedToolsPath)
if [ -z "$CACHED_RUNNER" ];then
    downloadAndExtractRunner
else
    sudo cp -a "$CACHED_RUNNER"  "/home/{{ .RunnerUsername }}/actions-runner"
    chown {{ .RunnerUsername }}:{{ .RunnerGroup }} -R "/home/{{ .RunnerUsername }}/actions-runner" || fail "failed to change owner"
fi
```

Reply:
Note: I have not tested the above :smile: , but something like that should work. You will need to rebuild your image with that cache in place. Let me know if this would be acceptable.

Reply:
lgtm (without having tested it)

Reply:
> So, we would like to add a override `cached_runner`, maybe at
> `/opt/cache/actions-runner/override`

Let's go with `/opt/cache/actions-runner/latest`

If it exists we just blindly trust that it is indeed the latest version and prefer it over whichever version is supplied by the instance.

We can also add a config flag to disable system updates, but that should come with a big warning for users to make sure they periodically generate new, updated images

Reply:
Let's make the disable_updates flag a config value, as there is little chance to want to enable updates for one pool but not another.

Reply:
How would you like to configure the `disable_updates` flag? I think this is business of the provider not garm, but all code that builds and serializes the userdata is vendored library code from garm.
That means we would either need to add more params to
```
func GetCloudConfig(bootstrapParams params.BootstrapInstance, tools github.RunnerApplicationDownload, runnerName string) (string, error) {
```
or extend `BootstrapInstance` by some new Userdata options? 
What do you recommend?

Reply:
This is indeed a provider scope change. The new providers reuse code from garm, but they don't have to. There is no reason why we can't have an alternate [GetCloudConfig()](https://github.com/cloudbase/garm-provider-openstack/blob/main/provider/spec.go#L201) that composes the cloud config with different values.

Reply:
I added https://github.com/cloudbase/garm-provider-openstack/pull/2 .. but it will only really work together if this is merged and the vendored garm is updated in `garm-provider-openstack`.